### PR TITLE
Custom derivative takers

### PR DESCRIPTION
**Pull request by isuruf**
- [x] Wait for https://github.com/inducer/pytential/pull/65

### CODE REVIEW

**Comment by inducer** [comment]
Retargeted to master after #24 was merged. Looks like there's a conflict. Could you take a look?

**Comment by inducer** [comment]
To cut down on noise in my email, I'm unsubscribing from this PR. When it next needs my attention, please @-mention me or hit the "request review" button. Otherwise, I may not see your messages in a timely manner.

**Comment by isuruf** [comment]
@inducer, this is ready for a review. pytential PR at inducer/pytential#65

**Comment by inducer** [comment]
@isuruf CI is failing here, could you take a look?

**Comment by isuruf** [comment]
This is ready now

**Comment by inducer** [comment]
@isuruf Do you think the thing I did in 803c344 is sane? That code will stick around on `main`, where it won't do much... (but it can get repurposed in the next situation like this).

**Comment by isuruf** [comment]
yep, looks good to me.

File: sumpy/assignment_collection.py [hunk 1]
```diff
@@ -191,6 +191,10 @@ def run_global_cse(self, extra_exprs=[]):
             assert isinstance(name, sym.Symbol)
             self.add_assignment(name.name, value)
 
+        for name, new_expr in zip(assign_names, new_assign_exprs):
+            del self.assignments[name]
+            self.assignments[name] = new_expr
```
Comment (inducer) [inline sumpy/assignment_collection.py:196]: IIUC, this is semantically a no-op. Add a comment on why it helps?

File: sumpy/tools.py [hunk 2]
```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
                 if (np.array(mi) >= np.array(other_mi)).all()),
             key=lambda other_mi: sum(self.mi_dist(mi, other_mi)))
 
+
+class LaplaceDerivativeTaker(MiDerivativeTaker):
```
Comment (inducer) [inline sumpy/tools.py:213]: Add a comment on what math this implements (and the complexity implications).

File: sumpy/tools.py [hunk 3]
```diff
@@ -84,28 +86,112 @@ def mi_power(vector, mi, evaluate=True):
...
+
+        .. math::
+
+            g^{(i)} (\xi) = \alpha^i f^{(i)} (\alpha \xi) = C' \cdot \alpha^i \cdot
+            \frac{1}{(\alpha x)^i} \quad (i > 0),
+
+        making them what you might call *scale-invariant*.
+
+        This derivative taker returns :math:`g^{(i)}(\xi) = \alpha^i f^{(i)}`
+        given :math:`f^{(0)}` as *expr* and :math:`\alpha` as *rscale*.
```
Comment (inducer) [inline sumpy/tools.py:167]:
```suggestion
        given :math:`f^{(0)}` as *expr* and :math:`\alpha` as :attr:`rscale`.
```

File: sumpy/tools.py [hunk 4]
```diff
@@ -84,28 +86,112 @@ def mi_power(vector, mi, evaluate=True):
...
 
-    def __init__(self, expr, var_list):
         assert isinstance(expr, sym.Basic)
         self.var_list = var_list
         empty_mi = (0,) * len(var_list)
         self.cache_by_mi = {empty_mi: expr}
+        self.rscale = rscale
+        self.sac = sac
+        self.dim = len(self.var_list)
+        self.orig_expr = expr
```
Comment (inducer) [inline sumpy/tools.py:177]: Maybe underscore these based on which are supposed to be exposed and not?

File: sumpy/tools.py [hunk 5]
```diff
@@ -84,28 +86,112 @@ def mi_power(vector, mi, evaluate=True):
...
+
+        This derivative taker returns :math:`g^{(i)}(\xi) = \alpha^i f^{(i)}`
+        given :math:`f^{(0)}` as *expr* and :math:`\alpha` as *rscale*.
+        """
 
-    def __init__(self, expr, var_list):
         assert isinstance(expr, sym.Basic)
         self.var_list = var_list
         empty_mi = (0,) * len(var_list)
         self.cache_by_mi = {empty_mi: expr}
```
Comment (inducer) [inline sumpy/tools.py:173]:
```suggestion
        zero_mi = (0,) * len(var_list)
        self.cache_by_mi = {zero_mi: expr}
```

File: sumpy/tools.py [hunk 6]
```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
...
+        r_expr = expr.xreplace({self.r**2: rsym**2})
+        self.is_radial = not any(r_expr.has(v) for v in var_list)
+        self.var_list_multiplied = [add_to_sac(sac, v * rscale) for v in var_list]
+
+    def diff(self, mi, q=0):
+        """
+        See [1] for the algorithm
+
+        [1]: Tausch, J., 2003. The fast multipole method for arbitrary Green's
+             functions. Contemporary Mathematics, 329, pp.307-314.
```
Comment (inducer) [inline sumpy/tools.py:288]: - https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#citations
- Use a symbolic name, like `Tausch2003`.

File: sumpy/tools.py [hunk 7]
```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
...
+        self.cache_by_mi_q = {(empty_mi, 0): expr}
+        self.r = sym.sqrt(sum(v**2 for v in var_list))
+        rsym = sym.Symbol("_r")
+        r_expr = expr.xreplace({self.r**2: rsym**2})
+        self.is_radial = not any(r_expr.has(v) for v in var_list)
+        self.var_list_multiplied = [add_to_sac(sac, v * rscale) for v in var_list]
+
+    def diff(self, mi, q=0):
+        """
+        See [1] for the algorithm
```
Comment (inducer) [inline sumpy/tools.py:285]: Talk a little bit about the properties and why this was chosen here.

File: sumpy/tools.py [hunk 8]
```diff
@@ -123,6 +209,159 @@ def get_closest_cached_mi(self, mi):
...
+            # See reference [1] in RadialDerivativeTaker.diff
+            k = (self.orig_expr * self.r).args[-1] / sym.I / self.r
+            expr = -(2*q - 1)/self.r**2 * self.diff(mi, q - 1)
+            expr += -k**2 / self.r * self.diff(mi, q - 2)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+MiDerivativeTakerWrapper = namedtuple("MiDerivativeTakerWrapper",
+                                      ["taker", "initial_mi"])
```
Comment (inducer) [inline sumpy/tools.py:363]: What's the purpose of this? (add a comment maybe)

**Review by inducer — COMMENTED** [review]
Here are some thoughts from a first look.

File: sumpy/tools.py (hunk 4 above)
Comment (isuruf) [reply sumpy/tools.py]: I guess none of them? Only `.diff` method is supposed to be used by the user.

File: sumpy/codegen.py [hunk 9]
```diff
@@ -178,9 +180,10 @@ def kill_trivial_assignments(assignments, retain_names=set()):
         r = substitute(expr, unsubst_rej)
         result.append((name, r))
 
-    logger.info(
-        "kill trivial assignments (plain): done, {nrej} assignments killed"
-        .format(nrej=len(rejected_assignments)))
+    nrej = len(rejected_assignments)
+    total_time = time() - start
+    logger.info(f"kill trivial assignments (plain): done, "
+                f"{nrej} assignments killed in {total_time:.3g} seconds")
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/codegen.py]: Done in 5fa2b2c3e7f31fed594da9119987d4521293def2

File: sumpy/assignment_collection.py (hunk 1 above)
Comment (isuruf) [reply sumpy/assignment_collection.py]: Done in ca4cedd09e50b208a4b636f59de947ccabfa9d41

File: sumpy/tools.py (hunk 8 above)
Comment (isuruf) [reply sumpy/tools.py:382]: Done in e93c6709f4aa41b3365f6549ad8126505962cdd8

File: sumpy/tools.py (hunk 6 above)
Comment (isuruf) [reply sumpy/tools.py]: Done in 113c9643507884756aa3dfa8782b46fbb87652e7

File: sumpy/tools.py (hunk 7 above)
Comment (isuruf) [reply sumpy/tools.py]: Done in 113c9643507884756aa3dfa8782b46fbb87652e7

File: sumpy/tools.py (hunk 2 above)
Comment (isuruf) [reply sumpy/tools.py:213]: Done in 113c9643507884756aa3dfa8782b46fbb87652e7

File: sumpy/tools.py [hunk 10]
```diff
@@ -84,28 +86,112 @@ def mi_power(vector, mi, evaluate=True):
     return result
 
 
-class MiDerivativeTaker:
+def add_to_sac(sac, expr):
+    import sumpy.symbolic as sym
+    if sac is None:
+        raise RuntimeError("")
```
Comment (inducer) [inline sumpy/tools.py:92]:
```suggestion
```

File: sumpy/expansion/local.py [hunk 11]
```diff
@@ -112,16 +114,55 @@ class VolumeTaylorLocalExpansionBase(LocalExpansionBase):
...
+        :arg avec: vector from source to center.
+        :arg bvec: vector from center to target. Not usually necessary,
+            except for line-Taylor expansion.
+        :arg sac: a symbolic assignment collection where temporary
+            expressions are stored.
+
+        :returns: a list of :mod:`sympy` expressions representing
+            the coefficients of the expansion.
+        """
+        from sumpy.tools import MiDerivativeTakerWrapper
```
Comment (inducer) [inline sumpy/expansion/local.py:130]: Add a comment comparing to `coefficients_from_source_vec` from the base class.

File: sumpy/expansion/local.py [hunk 12]
```diff
@@ -112,16 +114,55 @@ class VolumeTaylorLocalExpansionBase(LocalExpansionBase):
...
+
+        return result
+
+    def coefficients_from_source(self, kernel, avec, bvec, rscale, sac=None):
+        return self.coefficients_from_source_vec((kernel,), avec, bvec,
+                rscale=rscale, weights=(1,), sac=sac)
 
     def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
+        if not self.use_rscale:
+            rscale = 1
```
Comment (inducer) [inline sumpy/expansion/local.py:164]: ?

File: sumpy/expansion/local.py [hunk 13]
```diff
@@ -140,8 +181,9 @@ def evaluate(self, kernel, coeffs, bvec, rscale, sac=None):
...
     def translate_from(self, src_expansion, src_coeff_exprs, src_rscale,
             dvec, tgt_rscale, sac=None, _fast_version=True):
-        logger.info("building translation operator: %s(%d) -> %s(%d): start"
-                % (type(src_expansion).__name__,
+        logger.info("building translation operator for %s: %s(%d) -> %s(%d): start"
+                % (src_expansion.kernel,
+                    type(src_expansion).__name__,
                     src_expansion.order,
                     type(self).__name__,
                     self.order))
```
Comment (inducer) [inline sumpy/expansion/local.py:189]:
```suggestion
        logger.info("building translation operator for %s: %s(%d) -> %s(%d): start",
                    src_expansion.kernel,
                    type(src_expansion).__name__,
                    src_expansion.order,
                    type(self).__name__,
                    self.order)
```

File: sumpy/expansion/local.py [hunk 14]
```diff
@@ -348,6 +381,7 @@ def calc_max_mi(mis):
             # to compensate for differentiating which is done at the end.
             # This moves the two cancelling "rscales" closer to each other at
             # the end in the hope of helping rscale magnitude.
+            from sumpy.tools import MiDerivativeTaker
```
Comment (inducer) [inline sumpy/expansion/local.py:384]:
```suggestion
```

File: sumpy/expansion/multipole.py [hunk 15]
```diff
@@ -54,62 +53,57 @@ class VolumeTaylorMultipoleExpansionBase(MultipoleExpansionBase):
     Coefficients represent the terms in front of the kernel derivatives.
     """
 
-    def coefficients_from_source(self, kernel, avec, bvec, rscale, sac=None):
+    def coefficients_from_source_vec(self, kernels, avec, bvec, rscale, weights,
+            sac=None):
         from sumpy.kernel import KernelWrapper
```
Comment (inducer) [inline sumpy/expansion/multipole.py:58]: Compare to scalar version.

File: sumpy/expansion/local.py [hunk 16]
```diff
@@ -112,16 +114,55 @@ class VolumeTaylorLocalExpansionBase(LocalExpansionBase):
...
+                wrapper = MiDerivativeTakerWrapper(taker, mi)
+                mi_expr = knl.postprocess_at_source(wrapper, avec)
+                # By passing `rscale` to the derivative taker we are taking a scaled
+                # version of the derivative which is `expr.diff(mi)*rscale**sum(mi)`
+                # which might be implemented efficiently for kernels like Laplace.
+                # One caveat is that `postprocess_at_source` might take more
+                # derivatives which would multiply the expression by more `rscale`s
+                # than necessary as the derivative taker does not know about
+                # `postprocess_at_source`. This is corrected by dividing by `rscale`.
+                expr = mi_expr * add_to_sac(sac, weight / rscale ** pp_nderivatives)
```
Comment (inducer) [inline sumpy/expansion/local.py:153]: Note to self: understand this.

File: sumpy/expansion/local.py [hunk 17]
```diff
@@ -21,13 +21,15 @@
 """
```
Comment (inducer) [inline sumpy/expansion/local.py:21]: Note to self: OK other than `rscale` in the presence of derivatives.

File: sumpy/expansion/multipole.py [hunk 18]
```diff
@@ -54,62 +53,57 @@ class VolumeTaylorMultipoleExpansionBase(MultipoleExpansionBase):
...
+        expr_dict = kernel.get_derivative_transformation_at_target(expr_dict)
+        pp_nderivatives = single_valued(sum(mi) for mi in expr_dict.keys())
 
-        return kernel.postprocess_at_target(result, bvec)
+        result = []
+        for coeff, mi in zip(coeffs, self.get_coefficient_identifiers()):
+            wrapper = MiDerivativeTakerWrapper(taker, mi)
+            mi_expr = kernel.postprocess_at_target(wrapper, bvec)
+            # For details about this correction, see the explanation at
+            # VolumeTaylorLocalExpansionBase.coefficients_from_source
```
Comment (inducer) [inline sumpy/expansion/multipole.py:101]:
```suggestion
            # For details about this correction relating to rscale,
            # see the explanation at
            # VolumeTaylorLocalExpansionBase.coefficients_from_source
```

File: sumpy/expansion/multipole.py [hunk 19]
```diff
@@ -54,62 +53,57 @@ class VolumeTaylorMultipoleExpansionBase(MultipoleExpansionBase):
...
+            expr = coeff * mi_expr / rscale**pp_nderivatives
+            result.append(expr)
```
Comment (inducer) [inline sumpy/expansion/multipole.py:103]:
```suggestion
            result.append(coeff * mi_expr / rscale**pp_nderivatives)
```

File: sumpy/expansion/multipole.py [hunk 20]
```diff
@@ -22,13 +22,12 @@
```
Comment (inducer) [inline sumpy/expansion/multipole.py:22]: Note to self: OK other than `rscale` correction.

File: sumpy/kernel.py [hunk 21]
```diff
@@ -199,76 +197,19 @@ def get_expression(self, dist_vec):
...
-        making them what you might call *scale-invariant*. In this case, one
-        may set :attr:`has_efficient_scale_adjustment`. For these kernels only,
-        :meth:`adjust_for_kernel_scaling` provides a shortcut for scaled kernel
-        derivative computation. Given :math:`f^{(i)}` as *expr*, it directly
-        returns an expression for :math:`g^{(i)}`, where :math:`i` is given
-        as *nderivatives*.
-
-        :arg rscale: The scaling parameter :math:`\alpha` above.
+    def _diff(self, expr, vec, mi):
+        """Take the derivative of an expression or a MiDerivativeTakerWrapper
```
Comment (inducer) [inline sumpy/kernel.py:201]: > MiDerivativeTakerWrapper

markup?

File: sumpy/tools.py [hunk 22]
```diff
@@ -123,6 +209,178 @@ def get_closest_cached_mi(self, mi):
...
+
+
+# A :class:`MiDerivativeTaker` represents an expression and
+# lets you get derivatives of those. This wrapper represents
+# a derivative of an expression and is useful when then there's
+# no specialized derivative taker for the derivative expression,
+# but there's one for the expression. This avoids writing
+# specialized derivative takers for example the double layer
+# expression when there's already one for the single layer.
+MiDerivativeTakerWrapper = namedtuple("MiDerivativeTakerWrapper",
```
Comment (inducer) [inline sumpy/tools.py:380]: Better name?

Comment (inducer) [inline sumpy/tools.py:380]: Dataclass?

https://github.com/inducer/pytools/blob/d6f7c1f5bfd6a1cfdb2cc748d508d9b05ab603d0/setup.py#L44

File: sumpy/expansion/local.py (hunk 14 above)
Comment (isuruf) [reply sumpy/expansion/local.py:384]: This was imported early in the function before and it's moved to where it's used now

File: sumpy/expansion/local.py (hunk 11 above)
Comment (isuruf) [reply sumpy/expansion/local.py:132]: Done in 6e14519c616940f80d7d562754cbb721d8fdb27e

File: sumpy/tools.py (hunk 22 above)
Comment (isuruf) [reply sumpy/tools.py]: Done in fcc367db3f6c2233dcf0403a9855e0e91aeb1c48

File: sumpy/kernel.py (hunk 21 above)
Comment (isuruf) [reply sumpy/kernel.py]: This is removed now

File: sumpy/tools.py [hunk 23]
```diff
@@ -125,6 +208,195 @@ def get_closest_cached_mi(self, mi):
...
+            # See reference [Tausch2003] in RadialDerivativeTaker.diff
+            k = (self.orig_expr * self.r).args[-1] / sym.I / self.r
+            expr = -(2*q - 1)/self.r**2 * self.diff(mi, q - 1)
+            expr += -k**2 / self.r * self.diff(mi, q - 2)
+        self.cache_by_mi_q[(mi, q)] = expr
+        return expr
+
+
+@tag_dataclass
+class MiDerivativeTakerWrapper:
```
Comment (isuruf) [inline sumpy/tools.py:374]: @inducer, any thoughts on the name?

File: sumpy/tools.py [hunk 24]
```diff
@@ -125,6 +208,195 @@ def get_closest_cached_mi(self, mi):
...
+        # By passing `rscale` to the derivative taker we are taking a scaled
+        # version of the derivative which is `expr.diff(mi)*rscale**sum(mi)`
+        # which might be implemented efficiently for kernels like Laplace.
+        # One caveat is that we are taking more derivatives because of
+        # :attr:`derivative_transformation` which would multiply the
+        # expression by more `rscale`s than necessary. This is corrected by
+        # dividing by `rscale`.
+        result = 0
+        for extra_mi, coeff in self.derivative_transformation.items():
+            result += coeff * self.taker.diff(add_mi(mi, extra_mi))
```
Comment (inducer) [inline sumpy/tools.py:395]: `sum()` comprehension

File: sumpy/tools.py [hunk 25]
```diff
@@ -125,6 +208,195 @@ def get_closest_cached_mi(self, mi):
...
+class MiDerivativeTakerWrapper:
+    """A :class:`MiDerivativeTaker` represents an expression and
+    lets you get derivatives of those. This wrapper represents
+    a derivative of an expression and is useful when then there's
+    no specialized derivative taker for the derivative expression,
+    but there's one for the expression. This avoids writing
+    specialized derivative takers for example the double layer
+    expression when there's already one for the single layer."""
+    taker: MiDerivativeTaker
+    derivative_transformation: dict
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/tools.py:395]: Docs in 70b953e. Removed assumption in da38e72

File: sumpy/tools.py (hunk 24 above)
Comment (isuruf) [reply sumpy/tools.py]: Done in da38e72f

File: sumpy/tools.py [hunk 26]
```diff
@@ -86,28 +88,111 @@ def mi_power(vector, mi, evaluate=True):
...
+        return expr
+
+    if isinstance(expr, (sym.Number, sym.Symbol)):
+        return expr
+
+    name = sac.assign_unique("temp", expr)
+    return sym.Symbol(name)
+
+
+class ExprDerivativeTaker(object):
```
Comment (inducer) [inline sumpy/tools.py:103]:
```suggestion
class ExprDerivativeTaker(object):
    """Facilitates the efficient computation of (potentially) high-order
    derivatives of a given :mod:`sympy` expression *expr* while attempting
    to maximize the number of common subexpressions generated.
    
    This class defines the interface and realizes a baseline implementation.
    More specialized implementations may offer better efficiency for special
    cases.
    """
```

File: sumpy/tools.py [hunk 27]
```diff
@@ -30,6 +30,8 @@
 
  .. autoclass:: BlockIndexRanges
  .. autoclass:: MatrixBlockIndexRanges
+ .. autoclass:: ExprDerivativeTaker
```
Comment (inducer) [inline sumpy/tools.py:33]: What about the Laplace and radial ones?

File: sumpy/tools.py [hunk 28]
```diff
@@ -125,6 +210,208 @@ def get_closest_cached_mi(self, mi):
...
+    expression when there's already one for the single layer.
+
+    Attributes:
+        taker (ExprDerivativeTaker): A derivative taker for the base kernel
+
+        derivative_transformation (dict):
+            A dictionary mapping a derivative
+            multi-index to a coeffcient. The expression represented by this
+            derivative taker is the linear combination of the derivatives
+            of the expression for the base kernel.
```
Comment (inducer) [inline sumpy/tools.py:392]: Does this work? I'm not sure sumpy uses the numpy docstring convention.

File: sumpy/tools.py [hunk 29]
```diff
@@ -125,6 +210,208 @@ def get_closest_cached_mi(self, mi):
...
+
+@tag_dataclass
+class DifferentiatedExprDerivativeTaker:
+    """A :class:`ExprDerivativeTaker` represents an expression and
+    lets you get derivatives of those. This wrapper represents
+    a sum of derivatives of an expression and is useful when there's
+    no specialized derivative taker for the derivative expression,
+    but there's one for the expression. This avoids writing
+    specialized derivative takers for example the double layer
+    expression when there's already one for the single layer.
```
Comment (inducer) [inline sumpy/tools.py:383]:
```suggestion
    """Implements the :class:`ExprDerivativeTaker` interface
    for an expression that is itself a linear combination of
    derivatives of a base expression. To take the actual derivatives,
    it makes use of an underlying derivative taker *taker*.
```

Comment (inducer) [inline sumpy/tools.py:383]: General technical writing tip: never use contractions in non-colloquial writing.

**Review by inducer — COMMENTED** [review]
Thanks! A few documentation nits, otherwise good to go.

File: sumpy/tools.py (hunk 28 above)
Comment (isuruf) [reply sumpy/tools.py:397]: It's not ideal, but I couldn't find out how to add docstrings for attributes in a dataclass. Any suggestions?

Comment (inducer) [reply sumpy/tools.py:397]: Does `.. attribute::` not work?

File: sumpy/tools.py (hunk 27 above)
Comment (isuruf) [reply sumpy/tools.py:33]: Added

File: sumpy/tools.py (hunk 28 above)
Comment (isuruf) [reply sumpy/tools.py]: Yes, that works. Thanks. It still looks horrible though.
image

File: sumpy/tools.py [hunk 30]
```diff
@@ -125,6 +228,209 @@ def get_closest_cached_mi(self, mi):
...
+    for an expression that is itself a linear combination of
+    derivatives of a base expression. To take the actual derivatives,
+    it makes use of an underlying derivative taker *taker*.
+
+    .. attribute:: taker (ExprDerivativeTaker): A derivative taker for the
+            base kernel.
+    .. attribute:: derivative_transformation (dict): A dictionary mapping a
+            derivative multi-index to a coeffcient. The expression represented
+            by this derivative taker is the linear combination of the derivatives
+            of the expression for the base kernel.
```
Comment (inducer) [inline sumpy/tools.py:411]: That doesn't render right:
image

I'll FTFY.

Comment (inducer) [inline sumpy/tools.py:411]:
```suggestion
    .. attribute:: taker
        A :class:`ExprDerivativeTaker` for the base expression.
        
    .. attribute:: derivative_transformation
        A dictionary mapping a derivative multi-index to a coefficient.
        The expression represented by this derivative taker is the linear
        combination of the derivatives of the expression for the
        base expression.
```

**Comment by inducer** [comment]
Thanks!

File: sumpy/tools.py (hunk 30 above)
Comment (isuruf) [reply sumpy/tools.py]: Thanks for fixing this.

**Comment by inducer** [comment]
> Remove whitespace on blank line

Wow, that was fast. I tried to remove the space using Github's web editor---there was no more whitespace. Then I tried to do the same with a local checkout. Same result. Only then did I notice that you had already zapped the errant space. Thanks!

**Pull request merged**